code_review/16-09-20_s2s_util.py

```diff
@@ -0,0 +1,129 @@
+    - If a sample does not exist locally
+      - find its corresponding SRX number and all runs that belong to it,
+      - check if they all have the same layout, if not, crash
+      - see if we can download the runs from ena
+
+    output:
+        dict(
+            "GSM1234": {"layout": "PAIRED",
+                         "runs": ["SRR1234", "SRR4321"],
+                         "ena_fastq_http": {
+                            "SRR1234": [...],
+                            "SRR4321": None
+                            },
+                         "ena_fastq_ftp": ["..."],
+
+            "SRR5678": {"Layout": "SINGLE",
+                        "runs": ["SRR5678"],
+                        ena_fastq_http: None,
+                        ena_fastq_ftp: [...],
+            ...
+        )
+    """
+    # start with empty dictionary which we fill out later
+    sampledict = {sample: dict() for sample in samples}
+
+    # fill out the sampledict for the local samples, and store the public samples for later
+    public_samples = []
+    for sample in samples:
+        if os.path.exists(expand(f'{{fastq_dir}}/{sample}.{{fqsuffix}}.gz', **config)[0]):
+            sampledict[sample]["layout"] = "SINGLE"
+        elif all(os.path.exists(path) for path in expand(f'{{fastq_dir}}/{sample}_{{fqext}}.{{fqsuffix}}.gz', **config)):
+            sampledict[sample]["layout"] = "PAIRED"
+        elif sample.startswith(('GSM', 'SRX', 'SRR', 'ERR', 'DRR')):
+            public_samples.append(sample)
+        else:
+            raise ValueError(f"\nsample {sample} was not found..\n"
+                             f"We checked for SE file:\n"
+                             f"\t{config['fastq_dir']}/{sample}.{config['fqsuffix']}.gz \n"
+                             f"and for PE files:\n"
+                             f"\t{config['fastq_dir']}/{sample}_{config['fqext1']}.{config['fqsuffix']}.gz \n"
+                             f"\t{config['fastq_dir']}/{sample}_{config['fqext2']}.{config['fqsuffix']}.gz \n"
+                             f"and since the sample did not start with either GSM, SRX, SRR, ERR, and DRR we "
+                             f"couldn't find it online..\n")
+
+    if len(public_samples) == 0:
+        return sampledict
+
+    # only continue with public samples
+    db = pysradb.SRAweb()
+
+    # all samples that are on GEO (GSM numbers), must first be converted to SRA ids (SRX numbers)
+    geo_samples = [sample for sample in public_samples if sample.startswith("GSM")]
+
+    # in sample2clean we store the (potential GEO) sample name in a SRA compliant name
+    df = db.gsm_to_srx(geo_samples)
+    sample2clean = dict(zip(df.experiment_alias, df.experiment_accession))
+
+    # now add the already SRA compliant names with a reference to itself
+    sample2clean.update({sample: sample for sample in public_samples if sample not in geo_samples})
+
+    # check our samples on sra
+    df = db.sra_metadata(list(sample2clean.values()), detailed=True)
```

Comment:
this overwrites df, so you can rename it just as well to something like `sra_df`